extensions/ql-vscode/src/view/common/Telemetry.ts

```diff
@@ -0,0 +1,56 @@
+import * as React from "react";
+import { useState } from "react";
+import { vscode } from "../vscode-api";
+
+/**
+ * Wraps `React.useState` to output telemetry events whenever the value changes.
+ *
+ * The only catch is that when using a predicate to filter which values output telemetry,
+ * the setter only accepts a raw value, instead of a `(prevState: S) => S` function.
+ *
+ * @param defaultValue Default value to pass to React.useState
+ * @param telemetryAction Name of the telemetry event to output
+ * @param filterTelemetryOnValue If provided, only output telemetry events when the predicate returns true. If not provided always outputs telemetry.
+ * @returns A value and a setter function, just as if from `React.useState`
+ */
+export function useStateWithTelemetry<S>(
+  defaultValue: S | (() => S),
+  telemetryAction: string,
+): [S, React.Dispatch<React.SetStateAction<S>>];
+export function useStateWithTelemetry<S>(
+  defaultValue: S | (() => S),
+  telemetryAction: string,
+  filterTelemetryOnValue: (value: S) => boolean,
+): [S, React.Dispatch<S>];
+export function useStateWithTelemetry<S>(
```

Comment:
I realise I didn't explain why `useStateWithTelemetry` is so complicated. I wanted to add the ability to only output telemetry for certain value changes, which mostly means only when the user opens something and not when they close it. This isn't absolutely essential but it seems sensible to me to limit the telemetry to only the things we're most interested in. We can always change or removing the filtering in the future if we want to.

However, this led to problems accepting `React.SetStateAction<S>` which is `S | (prevValue: S) => S`. We need to pass the function to the `setter` and there's not a way I can find to get what the new value is. This means we can't pass the value to our filter function to know if we should emit telemetry or not.

The way around this that I've gone for is to have two different signatures, which shows that you can only use `React.SetStateAction<S>` when not using a filter function.

Do people think this is sensible and necessary, or is it over the top and just makes things complicated?

Reply:
I think we can simplify the call signature by using a different method of sending the telemetry event. Instead of sending the telemetry event when calling the setter, we can send the telemetry event when the value has actually been updated. That means we'd use `useEffect` on the value instead of wrapping `useState`.

However, that will not necessarily make the implementation simpler since we now need to deal with the initial render and the other values in the `useEffect` callback changing, so we'd probably need to use `useRef` to keep track of the previous value. Something like this:

```ts
export function useStateWithTelemetry<S>(
  defaultValue: S | (() => S),
  telemetryAction: string,
  filterTelemetryOnValue?: (value: S) => boolean,
): [S, React.Dispatch<React.SetStateAction<S>>] {
  const previousValue = useRef(defaultValue);
  const [value, setter] = useState<S>(defaultValue);

  useEffect(() => {
    if (value === previousValue.current) {
      return;
    }
    previousValue.current = value;

    if (filterTelemetryOnValue && !filterTelemetryOnValue(value)) {
      return;
    }

    sendTelemetry(telemetryAction);
  }, [telemetryAction, filterTelemetryOnValue, value, previousValue]);

  return [value, setter];
}
```

I haven't actually tested this, so I'm not sure whether this works in practice. Since we're now not actually using the setter, we could even simplify this to a different hook which doesn't need to wrap `useState` at all and you'd be able to use `useState` as-is. Something like this:

```ts
export function useTelemetryOnChange<S>(
  value: S,
  telemetryAction: string,
  filterTelemetryOnValue?: (value: S) => boolean,
): void {
  const previousValue = useRef(value);

  useEffect(() => {
    if (value === previousValue.current) {
      return;
    }
    previousValue.current = value;

    if (filterTelemetryOnValue && !filterTelemetryOnValue(value)) {
      return;
    }

    sendTelemetry(telemetryAction);
  }, [telemetryAction, filterTelemetryOnValue, value, previousValue]);
}
```

Again, I haven't actually tested this, but I believe the initial `value` that is passed in should ensure it's not called on first render.

Reply:
Thanks for taking a look here and giving ideas. I think both of these are great and should work. The second option where we're not wrapping `useState` is also very convenient and will make it easier to add in more options and use it more places. This will also make it easier to implement the debounce idea from earlier too, to avoid repeated events during filtering.